packages/core/src/fhir-deduplication/deduplicate-fhir.ts

```diff
@@ -26,353 +27,277 @@ import { deduplicateProcedures } from "./resources/procedure";
-  ["Practitioner", removeDanglingReferences],
-  ["Organization", removeDanglingReferences],
-]);
-
-const procedureFiltersMap = new Map<string, ResourceFilter>([
-  ["Practitioner", removeDanglingReferences],
-  ["Organization", removeDanglingReferences],
-]);
-
-const immunizationFiltersMap = new Map<string, ResourceFilter>([
-  ["Practitioner", removeDanglingReferences],
-  ["Organization", removeDanglingReferences],
-]);
-
-const resourceFiltersMap = new Map<string, Map<string, ResourceFilter>>([
-  ["AllergyIntolerance", allergiesFiltersMap],
-  ["Condition", conditionsFiltersMap],
-  ["Coverage", coveragesFiltersMap],
-  ["DiagnosticReport", diagReportFiltersMap],
-  ["Encounter", encounterFiltersMap],
-  ["Immunization", immunizationFiltersMap],
-  ["MedicationStatement", medicationRelatedFiltersMap],
-  ["MedicationRequest", medicationRelatedFiltersMap],
-  ["MedicationAdministration", medicationRelatedFiltersMap],
-  ["Composition", compositionFiltersMap],
-  ["Observation", observationFiltersMap],
-  ["Procedure", procedureFiltersMap],
-]);
-
 export function removeResourcesWithDanglingLinks(
-  entries: BundleEntry<Resource>[],
-  links: string[]
-) {
-  return entries.flatMap(entry => handleDanglingLinks(entry as Resource, links));
-}
+  resourceArrays: ExtractedFhirTypes,
+  danglingLinks: Set<string>
+): { updatedResourceArrays: ExtractedFhirTypes; deletedRefs: Set<string> } {
+  const deletedRefs = new Set<string>();
+  const updatedResourceArrays = initExtractedFhirTypes();
 
-function handleDanglingLinks(res: Resource, danglingLinks: string[]): Resource | [] {
-  if (res) {
-    let entry = res;
-    const filtersMap = resourceFiltersMap.get(res.resourceType);
-
-    if (filtersMap) {
-      for (const danglingLink of danglingLinks) {
-        const linkResourceType =
-          res.resourceType === "Composition" ? "all" : danglingLink.split("/")[0];
-        if (linkResourceType) {
-          const callbackFn = filtersMap.get(linkResourceType);
-          if (callbackFn) {
-            const result = callbackFn(entry, danglingLink);
-            if (result) entry = result;
-            else {
-              // removed resources become dangling links
-              danglingLinks.push(createRef(entry));
-              return [];
-            }
-          }
+  for (const [key, resources] of Object.entries(resourceArrays)) {
+    if (Array.isArray(resources)) {
+      const updatedResources: Resource[] = [];
+      for (const resource of resources) {
+        const result = removeDanglingReferences(resource, danglingLinks);
+        if (!result) {
+          const ref = createRef(resource);
+          deletedRefs.add(ref);
+        } else {
+          updatedResources.push(result);
         }
       }
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      updatedResourceArrays[key as keyof ExtractedFhirTypes] = updatedResources as any;
+    } else {
+      updatedResourceArrays.patient = resources;
```

Comment:
everything is the same as before except this line which is the BUG fix

Reply:
don't we want to be more explicit with this? i.e. check if `key === "patient"`? leaving this to an `else` seems like we're leaving it to potential bugs in the future

Reply:
patient is the only key that isnt an array according to the type definition. And if the definition changed and resources could be patient or something else, then it would throw a type error on build since we would be assigning something other than patient to an object of type patient

Reply:
type error on build is good